Add first draft of plugin architecture (#2)
Dekkonot/rojo: ARCHITECTURE.md

## The Plugin

As mentioned above, the Studio plugin uses the web server to receive patches that contain the changes, additions, and removals to the Studio datamodel. The patch does not actually point at instances directly but rather uses string IDs. The plugin has an InstanceMap that is a bidirectional map between instance IDs and Roblox instances. It lets us keep track of every instance that Rojo knows about.

To actually apply the patch to the Studio datamodel, the plugin has a Reconciler. The Reconciler uses the InstanceMap to find the instances mentioned by the patch, and then it attempts to modify/create/destroy instances as specified in the patch. It will also create and return a new patch that contains all of the operations that failed to apply (effectively creating a patch that if applied, would fix the failures), which is used to display warnings to the end user.

> The first time the plugin connects to the server is a bit of a special case. The server gives the plugin the current state, which is first used to hydrate the InstanceMap with all the initially known instances. Then, the Reconciler computes a patch by diffing the datamodel with the server state. This is known as the "catch up" patch, and that is what is shown in the Confirming page in the plugin. When that catch up patch is applied, the Studio datamodel should match the Rojo server and subsequent patches will be able to apply cleanly.

Reviewer: ```suggestion
The first time the plugin connects to the server, it receives the current state from the server. This is first used to populate the InstanceMap with all the initially known instances. Then, the Reconciler computes a patch by diffing the datamodel with the server state. This is known as the "catch up" patch, and that is what is shown in the Confirming page in the plugin. When that catch up patch is applied, the Studio datamodel should match the Rojo server and subsequent patches will be able to apply cleanly.
```

Active vs passive voice; it's not 'a bit of a special case', it is one. Also, `populate` is the right word here vs `hydrate` I think.

Author: The function is literally Reconciler:hydrate()
